### PR TITLE
added exec::reduce including a customization for static_thread_pool

### PR DESCRIPTION
**Pull request by muellan**
This adds 
- `exec::reduce`
- a parallelized specialization of `exec::reduce` for running on `exec::static_thread_pool`'s scheduler
- tests covering basic usage with different reduction operations, input value-range edge cases, ranges of non-fundamental types, different schedulers, etc.

This does not (yet) introduce a common customization point with `nvexec::reduce` mainly because this would require a common default reduction operation (`nvexec::reduce` uses `cub::Sum` by default while `exec::reduce` uses `std::plus<>`).
If this is desired then one should probably also provide an entire set of common reduction operations like min, max, sum, product, etc.

### CODE REVIEW

File: include/exec/static_thread_pool.hpp [hunk 1]
```diff
@@ -424,6 +422,263 @@ namespace exec {
...
+
+              sh_state.apply(reducer);
+
+              const bool is_last_thread = sh_state.finished_threads_.fetch_add(1)
+                                       == (total_threads - 1);
+              if (is_last_thread) {
+                sh_state.apply(finalize);
+              }
+            };
+          }
```
Comment (ericniebler) [inline include/exec/static_thread_pool.hpp:468]: We don't appear to be handling exceptions that may be thrown from the reduction function.

File: include/exec/reduce.hpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+      template <stdexec::sender _Sender, class _InitT, stdexec::__movable_value _RedOp>
+        requires(!stdexec::__tag_invocable_with_completion_scheduler<
+                  reduce_t,
+                  stdexec::set_value_t,
+                  _Sender,
+                  _InitT,
+                  _RedOp>)
+             && (!stdexec::tag_invocable<reduce_t, _Sender, _InitT, _RedOp>)
+      STDEXEC_DETAIL_CUDACC_HOST_DEVICE __sender<_Sender, _InitT, _RedOp>
```
Comment (ericniebler) [inline include/exec/reduce.hpp:185]: This should require that _RedOp is callable with (_InitT, _InitT), right?

File: include/exec/reduce.hpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            ((_Self&&) __self).__redop_};
+        }
+
+        template <stdexec::__decays_to<__t> _Self, class _Env>
+        friend auto tag_invoke(stdexec::get_completion_signatures_t, _Self&&, _Env&&)
+          -> stdexec::dependent_completion_signatures<_Env>;
+
+        template <stdexec::__decays_to<__t> _Self, class _Env>
+        friend auto tag_invoke(stdexec::get_completion_signatures_t, _Self&&, _Env&&)
+          -> stdexec::completion_signatures<stdexec::set_value_t(_InitT)>
```
Comment (ericniebler) [inline include/exec/reduce.hpp:130]: At this point, we have all the type information we need in order to do full type-checking. We can and should check that the sender can only complete one way, with a single value that is a range whose element type is the same as _InitT.

We also know whether the reduction function is potentially throwing, and if it is, to add `set_error_t(exception_ptr)` as a possible completion signature.

If you're feeling ambitious, you can see how an adaptor like `then` detects when the types don't match and tries to return a type that describes the error -- or propagates type errors from inner senders. If that's too much, don't sweat it. We can improve diagnostics in a separate PR.

File: include/exec/static_thread_pool.hpp [hunk 4]
```diff
@@ -424,6 +422,263 @@ namespace exec {
...
+        struct dummy_ {
+          using __t = T;
+        };
+
+        using inrange_t = //
+          stdexec::__decay_t<stdexec::__t<stdexec::__t< stdexec::__value_types_of_t<
+            Sender,
+            stdexec::env_of_t<Receiver>,
+            stdexec::__q<dummy_>,
+            stdexec::__q<dummy_>>>>>;
```
Comment (ericniebler) [inline include/exec/static_thread_pool.hpp:481]: I think this is equivalent:

```suggestion
        using inrange_t = //
          stdexec::__decay_t<
            stdexec::__value_types_of_t<
              Sender,
              stdexec::env_of_t<Receiver>,
              stdexec::__q<stdexec::__midentity>,
              stdexec::__q<stdexec::__midentity>>>;
```

File: include/exec/reduce.hpp [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "../stdexec/execution.hpp"
+
+#include <numeric>
+#include <ranges>
```
Comment (ericniebler) [inline include/exec/reduce.hpp:21]: I think it's ok for it to be an error to include exec/reduce.hpp if the stdlib doesn't support `<ranges>`, but the reduce test should simply not run in that case instead of causing a test failure.

File: include/exec/reduce.hpp [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+      struct __t {
+        using is_receiver = void;
+        using __id = __receiver;
+        __data* __op_;
+
+        template <
+          stdexec::__same_as<stdexec::set_value_t> _Tag,
+          class _Range,
+          class _Value = stdexec::range_value_t<_Range>>
```
Comment (ericniebler) [inline include/exec/reduce.hpp:45]: `std::ranges::range_value_t` ?

**Review by ericniebler — COMMENTED** [review]
Overall this looks really great, thanks Andre. A few nits, and needs some more type-checking.

File: include/exec/static_thread_pool.hpp [hunk 7]
```diff
@@ -424,6 +422,263 @@ namespace exec {
...
+            : sh_state_(sh_state) {
+            task_base::__execute = [](task_base* t, const std::uint32_t tid) noexcept {
+              auto& sh_state = *static_cast<reduction_task*>(t)->sh_state_;
+              auto total_threads = sh_state.num_threads_available();
+              auto& partial = sh_state.partials_[tid];
+
+              auto reducer = [&](auto& input) {
+                const auto [first, last] = even_share(std::ranges::size(input), tid, total_threads);
+                if (first != last) {
+                  auto begin = std::ranges::begin(input) + first;
```
[the remark this answers is not in the document]
Comment (muellan) [reply include/exec/static_thread_pool.hpp:443]: I know that I still need to add more type checking.
If it were up to me I would constrain it to random access ranges. On the other hand `std::reduce` requires only input iterators or forward iterators for the execution policy overloads. I don't know if anyone would actually expect `exec::reduce` to work with an input range.

File: include/exec/static_thread_pool.hpp (hunk 1 above)
Comment (muellan) [reply include/exec/static_thread_pool.hpp:469]: I know. I left exception handling out, just wanted stuff to work first. I'll add it.

File: include/exec/static_thread_pool.hpp [hunk 8]
```diff
@@ -424,6 +422,263 @@ namespace exec {
...
+              };
+
+              auto finalize = [&](auto&) {
+                auto result = std::reduce(
+                  std::ranges::begin(sh_state.partials_),
+                  std::ranges::end(sh_state.partials_),
+                  sh_state.init_,
+                  sh_state.redop_);
+                // deallocate
+                sh_state.partials_ = {};
```
[the remark this answers is not in the document]
Comment (muellan) [reply include/exec/static_thread_pool.hpp]: Can't believe I fell for this age-old trap.

File: include/exec/reduce.hpp (hunk 2 above)
Comment (muellan) [reply include/exec/reduce.hpp:185]: Right.

File: include/exec/reduce.hpp (hunk 3 above)
Comment (muellan) [reply include/exec/reduce.hpp:130]: I've seen the error handling in other sender adaptors. I think I should add it as part of this PR.

File: include/exec/reduce.hpp (hunk 5 above)
Comment (muellan) [reply include/exec/reduce.hpp:21]: agreed

File: include/exec/reduce.hpp (hunk 6 above)
Comment (muellan) [reply include/exec/reduce.hpp:45]: oops, that's a leftover I forgot to remove
